### PR TITLE
Create Basic Lambda - Sf emails to s3 exporter

### PR DESCRIPTION
**Pull request by david-pepper**
**What does this change?**
Contains the Handler, cfn.yaml and riff-raff.yaml files for as basic lambda in AWS that will be iterated upon with business logic and tests in subsequent PRs.

**Lambda purpose**
The intention of the Lambda is to regularly query for emails from Salesforce which as marked as 'Ready for Export to S3'.

The records returned by the query will be converted to Json and saved to S3

The lambda should be able to GET and PUT files in new S3 bucket (emails-from-sf/{Stage})

Scheduling and Error logging will follow in subsequent dedicated PRs

### CODE REVIEW

File: handlers/sf-emails-to-s3-exporter/cfn.yaml [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    Type: String
+    AllowedValues:
+      - PROD
+      - DEV
+    Default: DEV
+
+Mappings:
+  StageMap:
+    PROD:
+      SalesforceStage: PROD
```
Comment (kelvin-chappell) [inline handlers/sf-emails-to-s3-exporter/cfn.yaml:15]: Can just refer to the stage name for this value

File: handlers/sf-emails-to-s3-exporter/cfn.yaml [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+      SalesforceUsername: EmailsToS3APIUser
+      AppName: SFEmailsToS3
+      AppSecretsVersion: de769b81-3245-414a-acfb-a15c8ead92e5
+      SalesforceUserSecretsVersion: f0ef9b35-de4c-4fc7-a566-81aab194b9f4
+    DEV:
+      SalesforceStage: DEV
```
Comment (kelvin-chappell) [inline handlers/sf-emails-to-s3-exporter/cfn.yaml:21]: Can just refer to the stage name for this value

File: handlers/sf-emails-to-s3-exporter/cfn.yaml [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    Default: DEV
+
+Mappings:
+  StageMap:
+    PROD:
+      SalesforceStage: PROD
+      SalesforceUsername: EmailsToS3APIUser
+      AppName: SFEmailsToS3
+      AppSecretsVersion: de769b81-3245-414a-acfb-a15c8ead92e5
+      SalesforceUserSecretsVersion: f0ef9b35-de4c-4fc7-a566-81aab194b9f4
```
Comment (kelvin-chappell) [inline handlers/sf-emails-to-s3-exporter/cfn.yaml:19]: I wouldn't bother with the secrets version unless you think you might not want the latest version. It's easier to force the lambda to fetch new environment variables when there's a change in a secret.

File: handlers/sf-emails-to-s3-exporter/cfn.yaml [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    DEV:
+      SalesforceStage: DEV
+      SalesforceUsername: EmailsToS3APIUser
+      AppName: AwsConnectorSandbox
+      AppSecretsVersion: abaa595b-e6c8-4d13-81e0-2d02627536c7
+      SalesforceUserSecretsVersion: a73eadbb-c5f4-4249-bf90-31fa5d6738e3
```
Comment (kelvin-chappell) [inline handlers/sf-emails-to-s3-exporter/cfn.yaml:25]: I wouldn't bother with the secrets version unless you think you might not want the latest version. It's easier to force the lambda to fetch new environment variables when there's a change in a secret.

File: handlers/sf-emails-to-s3-exporter/cfn.yaml [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+      MemorySize: 512
+      Runtime: java8.al2
+      Timeout: 900
+      Environment:
+        Variables:
+          Stage: !Ref Stage
+          sfApiVersion: v46.0
+          sfAuthUrl:
+            !Sub
+            - '{{resolve:secretsmanager:${SalesforceStage}/Salesforce/ConnectedApp/${AppName}:SecretString:authUrl::${AppSecretsVersion}}}'
```
Comment (kelvin-chappell) [inline handlers/sf-emails-to-s3-exporter/cfn.yaml:49]:
```suggestion
            - '{{resolve:secretsmanager:${Stage}/Salesforce/ConnectedApp/${AppName}:SecretString:authUrl}}'
```

File: handlers/sf-emails-to-s3-exporter/cfn.yaml [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+            - SalesforceStage: !FindInMap [ StageMap, !Ref Stage, SalesforceStage ]
+              AppName: !FindInMap [ StageMap, !Ref Stage, AppName ]
+              AppSecretsVersion: !FindInMap [ StageMap, !Ref Stage, AppSecretsVersion ]
+          sfClientId:
+            !Sub
+            - '{{resolve:secretsmanager:${SalesforceStage}/Salesforce/ConnectedApp/${AppName}:SecretString:clientId::${AppSecretsVersion}}}'
```
Comment (kelvin-chappell) [inline handlers/sf-emails-to-s3-exporter/cfn.yaml:55]:
```suggestion
            - '{{resolve:secretsmanager:${Stage}/Salesforce/ConnectedApp/${AppName}:SecretString:clientId}}'
```

File: handlers/sf-emails-to-s3-exporter/cfn.yaml [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+            - SalesforceStage: !FindInMap [ StageMap, !Ref Stage, SalesforceStage ]
+              AppName: !FindInMap [ StageMap, !Ref Stage, AppName ]
+              AppSecretsVersion: !FindInMap [ StageMap, !Ref Stage, AppSecretsVersion ]
+          sfClientSecret:
+            !Sub
+            - '{{resolve:secretsmanager:${SalesforceStage}/Salesforce/ConnectedApp/${AppName}:SecretString:clientSecret::${AppSecretsVersion}}}'
```
Comment (kelvin-chappell) [inline handlers/sf-emails-to-s3-exporter/cfn.yaml:61]:
```suggestion
            - '{{resolve:secretsmanager:${Stage}/Salesforce/ConnectedApp/${AppName}:SecretString:clientSecret}}'
```

File: handlers/sf-emails-to-s3-exporter/cfn.yaml [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+            - SalesforceStage: !FindInMap [ StageMap, !Ref Stage, SalesforceStage ]
+              AppName: !FindInMap [ StageMap, !Ref Stage, AppName ]
+              AppSecretsVersion: !FindInMap [ StageMap, !Ref Stage, AppSecretsVersion ]
+          sfPassword:
+            !Sub
+            - '{{resolve:secretsmanager:${SalesforceStage}/Salesforce/User/${SalesforceUsername}:SecretString:sfPassword::${SalesforceUserSecretsVersion}}}'
```
Comment (kelvin-chappell) [inline handlers/sf-emails-to-s3-exporter/cfn.yaml:67]:
```suggestion
            - '{{resolve:secretsmanager:${Stage}/Salesforce/User/${SalesforceUsername}:SecretString:sfPassword}}'
```

File: handlers/sf-emails-to-s3-exporter/cfn.yaml [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+            - SalesforceStage: !FindInMap [ StageMap, !Ref Stage, SalesforceStage ]
+              SalesforceUserSecretsVersion: !FindInMap [ StageMap, !Ref Stage, SalesforceUserSecretsVersion ]
+              SalesforceUsername: !FindInMap [ StageMap, !Ref Stage, SalesforceUsername ]
+          sfToken:
+            !Sub
+            - '{{resolve:secretsmanager:${SalesforceStage}/Salesforce/User/${SalesforceUsername}:SecretString:sfToken::${SalesforceUserSecretsVersion}}}'
```
Comment (kelvin-chappell) [inline handlers/sf-emails-to-s3-exporter/cfn.yaml:73]:
```suggestion
            - '{{resolve:secretsmanager:${Stage}/Salesforce/User/${SalesforceUsername}:SecretString:sfToken}}'
```

File: handlers/sf-emails-to-s3-exporter/cfn.yaml [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+            - SalesforceStage: !FindInMap [ StageMap, !Ref Stage, SalesforceStage ]
+              SalesforceUserSecretsVersion: !FindInMap [ StageMap, !Ref Stage, SalesforceUserSecretsVersion ]
+              SalesforceUsername: !FindInMap [ StageMap, !Ref Stage, SalesforceUsername ]
+          sfUsername:
+            !Sub
+            - '{{resolve:secretsmanager:${SalesforceStage}/Salesforce/User/${SalesforceUsername}:SecretString:sfUsername::${SalesforceUserSecretsVersion}}}'
```
Comment (kelvin-chappell) [inline handlers/sf-emails-to-s3-exporter/cfn.yaml:79]:
```suggestion
            - '{{resolve:secretsmanager:${Stage}/Salesforce/User/${SalesforceUsername}:SecretString:sfUsername}}'
```

File: handlers/sf-emails-to-s3-exporter/project/build.properties [hunk 11]
```diff
@@ -0,0 +1 @@
+sbt.version=1.2.3
```
Comment (kelvin-chappell) [inline handlers/sf-emails-to-s3-exporter/project/build.properties:1]:
```suggestion
sbt.version=1.5.6
```

**Review by kelvin-chappell — COMMENTED** [review]
Looks good! Just a few comments

File: handlers/sf-emails-to-s3-exporter/cfn.yaml (hunk 2 above)
Comment (david-pepper) [reply handlers/sf-emails-to-s3-exporter/cfn.yaml:21]: Done

File: handlers/sf-emails-to-s3-exporter/cfn.yaml (hunk 1 above)
Comment (david-pepper) [reply handlers/sf-emails-to-s3-exporter/cfn.yaml:15]: Done

File: handlers/sf-emails-to-s3-exporter/cfn.yaml (hunk 3 above)
Comment (david-pepper) [reply handlers/sf-emails-to-s3-exporter/cfn.yaml:19]: So the lambda will work without adding these versions explicitly?

File: handlers/sf-emails-to-s3-exporter/cfn.yaml (hunk 5 above)
Comment (david-pepper) [reply handlers/sf-emails-to-s3-exporter/cfn.yaml]: I've removed SalesforceStage altogether from the file

File: handlers/sf-emails-to-s3-exporter/cfn.yaml (hunk 7 above)
Comment (david-pepper) [reply handlers/sf-emails-to-s3-exporter/cfn.yaml]: Done

File: handlers/sf-emails-to-s3-exporter/cfn.yaml (hunk 8 above)
Comment (david-pepper) [reply handlers/sf-emails-to-s3-exporter/cfn.yaml]: Done

File: handlers/sf-emails-to-s3-exporter/cfn.yaml (hunk 9 above)
Comment (david-pepper) [reply handlers/sf-emails-to-s3-exporter/cfn.yaml]: Done

File: handlers/sf-emails-to-s3-exporter/src/main/scala/com/gu/sf_emails_to_s3_exporter/Handler.scala [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+import com.typesafe.scalalogging.LazyLogging
+
+object Handler extends LazyLogging {
+
+  def main(args: Array[String]): Unit = {
+    handleRequest()
+  }
+
+  def handleRequest(): Unit = {
+
```
[the remark this answers is not in the document]
Comment (david-pepper) [reply handlers/sf-emails-to-s3-exporter/src/main/scala/com/gu/sf_emails_to_s3_exporter/Handler.scala:12]: Done

File: handlers/sf-emails-to-s3-exporter/cfn.yaml (hunk 3 above)
Comment (kelvin-chappell) [reply handlers/sf-emails-to-s3-exporter/cfn.yaml:18]: Yes.  The secrets version is quite tricky to find and by default the lambda will pick up the latest version of the secret when it's installed.

**Review by kelvin-chappell — APPROVED** [review]
👍

File: handlers/sf-emails-to-s3-exporter/cfn.yaml (hunk 3 above)
Comment (david-pepper) [reply handlers/sf-emails-to-s3-exporter/cfn.yaml]: I've removed the versions now

File: handlers/sf-emails-to-s3-exporter/cfn.yaml (hunk 4 above)
Comment (david-pepper) [reply handlers/sf-emails-to-s3-exporter/cfn.yaml]: Removed

File: build.sbt [hunk 13]
```diff
@@ -339,6 +339,11 @@ lazy val `soft-opt-in-consent-setter` = lambdaProject(
   Seq(awsSecretsManager, circe, circeParser, scalatest, scalajHttp, awsS3, simpleConfig) ++ logging
 ).dependsOn(`effects-s3`, `effects-cloudwatch`, `salesforce-core`)
 
+lazy val `sf-emails-to-s3-exporter` = lambdaProject(
+  "sf-emails-to-s3-exporter",
+  "Runs regularly to retrieve emails from Salesforce and save as json in S3",
+  Seq(awsSecretsManager, circe, circeParser, scalajHttp, awsS3))
```
Comment (kelvin-chappell) [inline build.sbt:345]: Are you actually using `awsSecretsManager` programmatically?  If it's just being used to find env vars at lambda install time, you don't need it as a dependency.
```suggestion
  Seq(circe, circeParser, scalajHttp, awsS3))
```

Comment (david-pepper) [reply build.sbt]: Done

**Pull request merged**